### DISCUSSION TITLE
Support for other AWS Resources

### DISCUSSION THREAD

**Discussion by markussiebert** [renovatebot/renovate#24923] [Suggest an Idea]
### What would you like Renovate to be able to do?

If you deploy a lot of infrastructure with AWS you may face several versioning problems. I want to give you some examples:

1. RDS (relational databases)

If you deploy databases you have to specify engine versions of your database instance. Thats fine ... you lookup and choose that one you like and run your infrastructure as code magic. But than you have to take care of this instance and update it periodically. There you have 2 possibilities:

a) DIY
you have to keep track of the versions and update your infrastructure code once in a while ... sound's like what renovate does.

b) let AWS handle it
yes, AWS can perform major and minor version upgrades, but now and then, you have to do changes to your database, something like increase storage, or change some configuration. But then, your scripts updating the infrastructure will fail if you don't update the version of the database to the actual used version. That's the point where managing this with renovate would be better.

Other services that rely on versions and have the same:

2. MSK (managed kafka)
3. Redis
4. Opensearch
5. ... even more to come ...

### If you have any ideas on how this should be implemented, please tell us here.

Maybe we could create a datasource for each of these services?

Would like to hear your opinion on this @rarkins 

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by rarkins** [comment]
This sounds like a good idea. Do you know if AWS exposes any sort of API so that Renovate could query to learn e.g. the supported versions of Postgres in RDS?

**Comment by markussiebert** [comment]
I have to look into every Service, if they have something like apis, but for RDS we have the DescribeDbEngineVersions Command.

https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-rds/classes/describedbengineversionscommand.html

I think we should start with the aws-rds-datasource, because I think this is the most important service (at least for me).

**Comment by HonkingGoose** [comment]
The Amazon RDS docs cover a `DescribeDBEngineVersions` [^docs] command, which does what you want, I think. 

> Returns a list of the available DB engines.

Sample response:

```xml
<DescribeDBEngineVersionsResponse xmlns="http://rds.amazonaws.com/doc/2014-09-01/">
  <DescribeDBEngineVersionsResult>
    <DBEngineVersions>
      <DBEngineVersion>
        <Engine>mysql</Engine>
        <DBParameterGroupFamily>mysql5.1</DBParameterGroupFamily>
        <DBEngineDescription>MySQL Community Edition</DBEngineDescription>
        <EngineVersion>5.1.57</EngineVersion>
        <DBEngineVersionDescription>MySQL 5.1.57</DBEngineVersionDescription>
      </DBEngineVersion>
      <DBEngineVersion>
        <Engine>mysql</Engine>
        <DBParameterGroupFamily>mysql5.1</DBParameterGroupFamily>
        <DBEngineDescription>MySQL Community Edition</DBEngineDescription>
        <EngineVersion>5.1.61</EngineVersion>
        <DBEngineVersionDescription>MySQL 5.1.61</DBEngineVersionDescription>
      </DBEngineVersion>
      <DBEngineVersion>
        <Engine>mysql</Engine>
        <DBParameterGroupFamily>mysql5.1</DBParameterGroupFamily>
        <DBEngineDescription>MySQL Community Edition</DBEngineDescription>
        <EngineVersion>5.1.62</EngineVersion>
        <DBEngineVersionDescription>MySQL 5.1.62</DBEngineVersionDescription>
      </DBEngineVersion>
  </DescribeDBEngineVersionsResult>
  <ResponseMetadata>
    <RequestId>b74d2635-b98c-11d3-fbc7-5c0aad74da7c</RequestId>
  </ResponseMetadata>
</DescribeDBEngineVersionsResponse>
```

[^docs]: https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_DescribeDBEngineVersions.html

**Comment by rarkins** [comment]
Could be nice if we have a single datasource - maybe even called just `aws` which has a known set of lookup names it supports, e.g. `rds-postgres`, etc.

**Comment by antoinetielbeke** [comment]
This would be a great addition! I think Elasticache (Redis) will be hard since AWS forces you to let them manage the minor version starting 6.x. Even the AWS CLI only returns 6.0 and 6.2.
```
        {
            "Engine": "redis",
            "EngineVersion": "5.0.6",
            "CacheParameterGroupFamily": "redis5.0",
            "CacheEngineDescription": "Redis",
            "CacheEngineVersionDescription": "redis version 5.0.6"
        },
        {
            "Engine": "redis",
            "EngineVersion": "6.0",
            "CacheParameterGroupFamily": "redis6.x",
            "CacheEngineDescription": "Redis",
            "CacheEngineVersionDescription": "redis version 6.0.5"
        },
        {
            "Engine": "redis",
            "EngineVersion": "6.2",
            "CacheParameterGroupFamily": "redis6.x",
            "CacheEngineDescription": "Redis",
            "CacheEngineVersionDescription": "redis version 6.2.5"
        }
```
I agree with @markussiebert that RDS would be the highest on my (personal) priority list.

**Comment by markussiebert** [comment]
I started last week to play around with the Apis and the renovate Code. But
i have some questions we should discuss ...

antoinetielbeke ***@***.***> schrieb am Di., 10. Mai 2022,
04:33:

**Comment by danports** [comment]
@markussiebert Were you able to make any progress with this? I may take a crack at this, for RDS at least, unless you've already got something on the way.

**Comment by danports** [comment]
> Could be nice if we have a single datasource - maybe even called just `aws` which has a known set of lookup names it supports, e.g. `rds-postgres`, etc.

@rarkins How would you see this working given that the `aws-machine-image` datasource customizes its `defaultConfig` fairly extensively? It seems to me that both `defaultConfig` and `defaultVersioning` would need to be methods rather than fields on the datasource interface for a single `aws` datasource to work as you've described. My lazy implementation approach here would just be to copy & paste the `aws-machine-image` datasource to a new `aws-rds` datasource and swap out the API calls and a few other things.

**Comment by rarkins** [comment]
@danports that's ok with me if combining them is too difficult

**Comment by danports** [comment]
The `aws-rds` datasource is available now! 🎉 However, I do not have any plans at present to implement any other AWS datasources, so I think we can leave this open to consider other AWS product-specific datasources or a more general `aws` datasource.
